code/datums/martial/worldshaker.dm

```diff
@@ -0,0 +1,563 @@
+		drop()
+		return 
+	if(iscarbon(tossed))
+		var/mob/living/carbon/tossedliving = thrown[1]
+		var/obj/item/bodypart/limb_to_hit = tossedliving.get_bodypart(user.zone_selected)
+		if(!tossedliving.buckled)
+			return
+		grab(user, tossedliving, throwdam) // Apply damage
+		for(var/obj/structure/bed/grip/F in view(2, user))
+			F.Destroy()
+		if(!limb_to_hit)
+			limb_to_hit = tossedliving.get_bodypart(BODY_ZONE_CHEST)
+	user.visible_message(span_warning("[user] throws [tossed]!"))
+	for(var/i = 1 to target_dist)
+		if(QDELETED(tossed))//if the thrown person is deleted,
+			return
+		var/dir_to_target = get_dir(get_turf(tossed), D) //vars that let the thing be thrown while moving similar to things thrown normally
+		var/turf/T = get_step(get_turf(tossed), dir_to_target)
+		if(objdam <= 0)// or total damage has run out, end the throw
+			playsound(T, 'sound/effects/gravhit.ogg', 40, TRUE, 5)
+			return
+		if(T.density) // crash into a wall and damage everything flying towards it before stopping 
+			for(var/mob/living/S in thrown)
+				grab(user, S, slamdam) 
+				S.Knockdown(1.5 SECONDS)
+				S.Immobilize(1.5 SECONDS)
+				if(isanimal(S) && S.stat == DEAD)
+					S.gib()	
+			drop()
+			return
+		for(var/obj/Z in T.contents) // crash into something solid and damage it along with thrown objects that hit it
+			if(Z.density) // If the thing is solid and anchored like a window or grille or table it hurts people thrown that crash into it too
+				for(var/mob/living/L in thrown) 
+					grab(user, L, slamdam) 
+					L.Knockdown(1.5 SECONDS)
+					L.Immobilize(1.5 SECONDS)
+					if(isanimal(L) && L.stat == DEAD)
+						L.gib()
+					if(istype(Z, /obj/machinery/disposal/bin)) // dumpster living things tossed into the trash
+						var/obj/machinery/disposal/bin/dumpster = D
+						L.forceMove(Z)
+						Z.visible_message(span_warning("[L] is thrown down the trash chute!"))
+						dumpster.do_flush()
+						drop()
+						return
+				var/reduction = Z.obj_integrity
+				Z.take_damage(objdam)
+				objdam -= reduction
+				if(Z.density)
+					playsound(Z, 'sound/effects/gravhit.ogg', 40, TRUE, 5)
+					drop()
+					return // if the solid thing we hit doesnt break then the thrown thing is stopped
+		for(var/mob/living/hit in T.contents) // if the thrown mass hits a person then they get tossed and hurt too along with people in the thrown mass
+			if(user != hit)
+				grab(user, hit, slamdam) 
+				hit.Knockdown(1.5 SECONDS) 
+				for(var/mob/living/L in thrown)
+					grab(user, L, slamdam) 
+					L.Knockdown(1 SECONDS) 
+				thrown |= hit
+		if(T) // if the next tile wont stop the thrown mass from continuing
+			for(var/mob/living/S in thrown)
+				S.Knockdown(1.5 SECONDS)
+				S.Immobilize(1.5 SECONDS)
+			for(var/atom/movable/K in thrown) // to make the mess of things that's being thrown almost look like a normal throw
+				K.SpinAnimation(0.2 SECONDS, 1) 
+				sleep(0.001 SECONDS)
```

Comment:
m

Reply:
like, i get it, sleep is bad, but rewriting this to instead be a recursive proc would be a nightmare